File: frontend/src/types/lp.ts
/**
 * Limited Partner (LP) TypeScript interfaces
 * Matches the backend LP model and API responses
 */

export interface LP {
  id: string;
  name: string;

  // Organization details
  type?: 'Individual' | 'Family Office' | 'Institution' | 'Corporate' | 'Foundation' | 'Government' | 'Other';
  description?: string;
  website?: string;

  // Contact information
  primary_contact_name?: string;
  primary_contact_email?: string;
  primary_contact_phone?: string;
  location?: string; // City, Country

  // Investment details
  total_committed_capital_raw?: string; // e.g., "$50M"
  total_committed_capital?: number; // e.g., 50000000
  investment_focus?: string; // e.g., "Technology, Healthcare"
  first_investment_year?: number;

  // Relationship tracking
  relationship_status?: 'Active' | 'Prospective' | 'Inactive' | 'Former';
  tier?: 'Tier 1' | 'Tier 2' | 'Tier 3'; // Investment size tiers

  // Timestamps
  created_at?: string;
  updated_at?: string;
}

export interface LPSearchParams {
  search?: string;
  type?: string;
  location?: string;
  relationship_status?: string;
  tier?: string;
  min_commitment?: number;
  max_commitment?: number;
  min_investment_year?: number;
  max_investment_year?: number;
  sort_by?: 'name' | 'total_committed_capital' | 'first_investment_year' | 'created_at';
  order?: 'asc' | 'desc';
  limit?: number;
  offset?: number;
}

export interface LPListResponse {
  lps: LP[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface LPStatistics {
  total_lps: number;
  total_committed_capital: number;
  avg_commitment: number;
  type_breakdown: Array<{
    type: string;
    count: number;
  }>;
  tier_breakdown: Array<{
    tier: string;
    count: number;
  }>;
  status_breakdown: Array<{
    status: string;
    count: number;
  }>;
}

export interface LPTypesResponse {
  types: string[];
}

// LP Types constants
export const LP_TYPES = [
  'Individual',
  'Family Office',
  'Institution',
  'Corporate',
  'Foundation',
  'Government',
  'Other'
] as const;

export const RELATIONSHIP_STATUSES = [
  'Active',
  'Prospective',
  'Inactive',
  'Former'
] as const;

export const TIERS = [
  'Tier 1',
  'Tier 2',
  'Tier 3'
] as const;

// Fund Commitment interface for LP investments
export interface FundCommitment {
  id: string;
  lp_id: string;
  fund_id: string;
  fund_name?: string; // Denormalized for display
  fund_strategy?: string; // Denormalized for display
  commitment_amount_raw?: string; // e.g., "$10M"
  commitment_amount?: number; // e.g., 10000000
  commitment_date?: string; // ISO date string
  capital_called?: number; // Amount of capital called to date
  capital_called_raw?: string; // Formatted display
  percent_called?: number; // Calculated: capital_called / commitment_amount * 100
  status?: 'Active' | 'Fully Called' | 'Cancelled';
  notes?: string;
  created_at?: string;
  updated_at?: string;
}

// LP Holding interface for portfolio holdings
export interface LPHolding {
  id: string;
  fund_id?: string;
  fund_name: string;
  vintage?: number; // Fund vintage year

  // Capital flows
  capital_committed?: number;
  capital_committed_raw?: string;

  capital_contributed?: number;
  capital_contributed_raw?: string;

  capital_distributed?: number;
  capital_distributed_raw?: string;

  market_value?: number;
  market_value_raw?: string;

  // Performance
  inception_irr?: number; // Percentage

  // Optional LP linkage
  lp_id?: string;
  lp_name?: string;

  // Timestamps
  created_at?: string;
  updated_at?: string;
}

export interface LPHoldingsListResponse {
  holdings: LPHolding[];
  total: number;
  limit: number;
  offset: number;
}

export interface LPHoldingsStats {
  total_capital_committed: number;
  total_capital_contributed: number;
  total_capital_distributed: number;
  total_market_value: number;
  average_irr: number;
  count: number;
  by_vintage: Record<string, number>;
}

export interface LPHoldingsSearchParams {
  lp_id?: string;
  vintage?: number;
  min_value?: number;
  max_value?: number;
  search?: string;
  sort_by?: 'vintage' | 'fund_name' | 'market_value' | 'inception_irr' | 'capital_committed' | 'capital_contributed';
  sort_order?: 'asc' | 'desc';
  limit?: number;
  offset?: number;
}

File: backend/schemas/lp.py
"""
Pydantic schemas for LP (Limited Partner) API endpoints
"""

from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime


class LPBase(BaseModel):
    """Base LP schema with common fields"""
    name: str = Field(..., description="LP name")

    # Organization details
    type: Optional[str] = Field(None, description="LP type (Individual, Family Office, Institution, Corporate, Foundation, Government, Other)")
    description: Optional[str] = Field(None, description="LP description")
    website: Optional[str] = Field(None, description="LP website URL")

    # Contact information
    primary_contact_name: Optional[str] = Field(None, description="Primary contact person name")
    primary_contact_email: Optional[str] = Field(None, description="Primary contact email")
    primary_contact_phone: Optional[str] = Field(None, description="Primary contact phone")
    location: Optional[str] = Field(None, description="LP location (City, Country)")

    # Investment details
    total_committed_capital_raw: Optional[str] = Field(None, description="Total committed capital as display string (e.g., '$50M')")
    total_committed_capital: Optional[float] = Field(None, description="Total committed capital as float (e.g., 50000000.0)")
    investment_focus: Optional[str] = Field(None, description="Investment focus areas (e.g., 'Technology, Healthcare')")
    first_investment_year: Optional[int] = Field(None, description="Year of first investment")

    # Relationship tracking
    relationship_status: Optional[str] = Field(None, description="Relationship status (Active, Prospective, Inactive, Former)")
    tier: Optional[str] = Field(None, description="Investment tier (Tier 1, Tier 2, Tier 3)")


class LPCreate(LPBase):
    """Schema for creating a new LP"""
    pass


class LPUpdate(BaseModel):
    """Schema for updating an LP (all fields optional)"""
    name: Optional[str] = None
    type: Optional[str] = None
    description: Optional[str] = None
    website: Optional[str] = None
    primary_contact_name: Optional[str] = None
    primary_contact_email: Optional[str] = None
    primary_contact_phone: Optional[str] = None
    location: Optional[str] = None
    total_committed_capital_raw: Optional[str] = None
    total_committed_capital: Optional[float] = None
    investment_focus: Optional[str] = None
    first_investment_year: Optional[int] = None
    relationship_status: Optional[str] = None
    tier: Optional[str] = None


class LPResponse(LPBase):
    """Schema for LP responses (includes ID and timestamps)"""
    id: str = Field(..., description="Unique LP ID")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")

    class Config:
        from_attributes = True  # Enable ORM mode for SQLAlchemy models


class LPListResponse(BaseModel):
    """Schema for paginated list of LPs"""
    lps: List[LPResponse] = Field(..., description="List of LPs")
    total: int = Field(..., description="Total number of LPs matching query")
    page: int = Field(..., description="Current page number (1-indexed)")
    page_size: int = Field(..., description="Number of items per page")
    total_pages: int = Field(..., description="Total number of pages")

    class Config:
        from_attributes = True


class LPSearchFilters(BaseModel):
    """Schema for LP search and filter parameters"""
    search: Optional[str] = Field(None, description="Search query for LP name")
    type: Optional[str] = Field(None, description="Filter by LP type")
    location: Optional[str] = Field(None, description="Filter by location")
    relationship_status: Optional[str] = Field(None, description="Filter by relationship status")
    tier: Optional[str] = Field(None, description="Filter by tier")
    min_commitment: Optional[float] = Field(None, description="Minimum commitment filter")
    max_commitment: Optional[float] = Field(None, description="Maximum commitment filter")
    min_investment_year: Optional[int] = Field(None, description="Minimum first investment year")
    max_investment_year: Optional[int] = Field(None, description="Maximum first investment year")

    # Pagination
    page: int = Field(1, ge=1, description="Page number (1-indexed)")
    page_size: int = Field(20, ge=1, le=100, description="Items per page (max 100)")

    # Sorting
    sort_by: Optional[str] = Field("name", description="Field to sort by (name, total_committed_capital, first_investment_year, created_at)")
    sort_order: Optional[str] = Field("asc", description="Sort order: 'asc' or 'desc'")


class LPFundCommitmentBase(BaseModel):
    """Base LP-Fund commitment schema"""
    lp_id: str = Field(..., description="LP ID")
    fund_id: str = Field(..., description="Fund ID")
    commitment_amount_raw: Optional[str] = Field(None, description="Commitment amount as display string (e.g., '$10M')")
    commitment_amount: Optional[float] = Field(None, description="Commitment amount as float (e.g., 10000000.0)")
    commitment_date: Optional[datetime] = Field(None, description="Date of commitment")
    capital_called_raw: Optional[str] = Field(None, description="Capital called as display string (e.g., '$7M')")
    capital_called: Optional[float] = Field(None, description="Capital called as float (e.g., 7000000.0)")
    notes: Optional[str] = Field(None, description="Additional notes")


class LPFundCommitmentCreate(LPFundCommitmentBase):
    """Schema for creating a new LP-Fund commitment"""
    pass


class LPFundCommitmentUpdate(BaseModel):
    """Schema for updating an LP-Fund commitment (all fields optional)"""
    commitment_amount_raw: Optional[str] = None
    commitment_amount: Optional[float] = None
    commitment_date: Optional[datetime] = None
    capital_called_raw: Optional[str] = None
    capital_called: Optional[float] = None
    notes: Optional[str] = None


class LPFundCommitmentResponse(LPFundCommitmentBase):
    """Schema for LP-Fund commitment responses (includes ID and timestamps)"""
    id: str = Field(..., description="Unique commitment ID")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")

    class Config:
        from_attributes = True


class LPTypesResponse(BaseModel):
    """Schema for LP types metadata"""
    types: List[str] = Field(..., description="List of available LP types")


class LPStatistics(BaseModel):
    """Schema for LP statistics"""
    total_lps: int = Field(..., description="Total number of LPs")
    total_committed_capital: float = Field(..., description="Total committed capital across all LPs")
    avg_commitment: float = Field(..., description="Average commitment per LP")
    type_breakdown: List[dict] = Field(..., description="Breakdown by LP type")
    tier_breakdown: List[dict] = Field(..., description="Breakdown by tier")
    status_breakdown: List[dict] = Field(..., description="Breakdown by relationship status")
